### PR TITLE
(#2) Add iOS sdks for epson and generic printers

### PR DESCRIPTION
**Pull request by mdsiddiq-till**
## Description

Add iOS sdks for epson and generic printers
It switches based on `brand` argument

## Related Issues

Issue: #XXX

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## Screenshots (if appropriate)

### CODE REVIEW

File: ios/PrinterSDK/GenericPrinterSDK.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+- (void)disconnect;
+
+- (void)setPrintWidth:(NSInteger)width;
+
+- (void)printText:(NSString*)text;
+- (void)printTextImage:(NSString*)text;
+- (void)sendHex:(NSString*)hex;
+
+- (void)printCodeBar:(NSString*)text type:(CodeBarType)type;
+- (void)printQrCode:(NSString*)text;
```
Comment (lakshmaji-till) [inline ios/PrinterSDK/GenericPrinterSDK.h:61]: do we need these ? printQrCode, etc. they were covered in xml lib!

File: ios/adapters/epson/USBPrinterEpsonAdapter.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
+//
+//  USBPrinterEpsonAdapter.h
+//
+//  Created by Till POS on 14/09/21.
+//
+
+
+#ifndef USBPrinterEpsonAdapter_h
```
Comment (lakshmaji-till) [inline ios/adapters/epson/USBPrinterEpsonAdapter.h:8]: is this commented code ?

File: src/index.ts [hunk 3]
```diff
@@ -4,26 +4,34 @@ const RNUSBPrinter = NativeModules.RNUSBPrinter;
...
   beep?: boolean;
   cut?: boolean;
   tailingLine?: boolean;
   encoding?: string;
 }
 
-export interface IUSBPrinter {
+export interface IBasePrinter {
   device_name: string;
+  brand?: PrinterBrand;
```
Comment (lakshmaji-till) [inline src/index.ts:22]: what happens if we provided brand input ?

File: ios/RNBLEPrinter.m [hunk 4]
```diff
@@ -8,17 +8,12 @@
 #import <Foundation/Foundation.h>
 
 #import "RNBLEPrinter.h"
-#import "PrinterSDK.h"
+#import "adapters/epson/BLEPrinterEpsonAdapter.h"
+#import "adapters/general/BLEPrinterGenericAdapter.h"
```
Comment (snehithv) [inline ios/RNBLEPrinter.m:12]: you mean `adapters/generic/BLEPrinterGenericAdapter.h`

File: ios/adapters/epson/BLEPrinterEpsonAdapter.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+//  BLEPrinterEpsonAdapter.h
+//
+//  Created by Till POS on 14/09/21.
+//
+
+#pragma once
+#ifndef BLEPrinterEpsonAdapter_h
+#define BLEPrinterEpsonAdapter_h
+
+#if __has_include("RCTBridgeModule.h")
```
Comment (snehithv) [inline ios/adapters/epson/BLEPrinterEpsonAdapter.h:11]: what is this check for?

File: src/index.ts (hunk 3 above)
Comment (snehithv) [reply src/index.ts:22]: we switch between using epsonSDK vs GenericSDK

File: ios/PrinterSDK/GenericPrinterSDK.h (hunk 1 above)
Comment (mdsiddiq-till) [reply ios/PrinterSDK/GenericPrinterSDK.h:61]: Removing from here won't have any effect.
It's just header file.

File: ios/RNBLEPrinter.m (hunk 4 above)
Comment (mdsiddiq-till) [reply ios/RNBLEPrinter.m:12]: generic or general?
Ok. Will change them to generic.

File: ios/adapters/epson/BLEPrinterEpsonAdapter.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
+//
+//  BLEPrinterEpsonAdapter.h
+//
+//  Created by Till POS on 14/09/21.
+//
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (mdsiddiq-till) [reply ios/adapters/epson/BLEPrinterEpsonAdapter.h:7]: _It means that imported classes will be loaded only once when app runs._

File: ios/adapters/epson/BLEPrinterEpsonAdapter.h (hunk 5 above)
Comment (mdsiddiq-till) [reply ios/adapters/epson/BLEPrinterEpsonAdapter.h:11]: Checks for the dependency if exists and then imports.
Like we did before .. similar code

File: ios/adapters/epson/USBPrinterEpsonAdapter.h (hunk 2 above)
Comment (mdsiddiq-till) [reply ios/adapters/epson/USBPrinterEpsonAdapter.h:8]: Not commented

File: ios/adapters/general/NetPrinterGenericAdapter.m [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+- (void) connectAndSend:(NSString *)host
+               withPort:(nonnull NSNumber *)port
+           printRawData:(NSString *)text
+                success:(RCTResponseSenderBlock)successCallback
+                   fail:(RCTResponseSenderBlock)errorCallback {
+    @try {
+        BOOL isConnectSuccess = [[PrinterSDK defaultPrinterSDK] connectIP:host];
+        !isConnectSuccess ? [NSException raise:@"Invalid connection" format:@"Can't connect to printer %@", host] : nil;
+        [[PrinterSDK defaultPrinterSDK] sendHex:text];
```
[the remark this answers is not in the document]
Comment (mdsiddiq-till) [reply ios/adapters/general/NetPrinterGenericAdapter.m]: Handled it by doing hex transform

File: android/src/main/java/com/pinmi/react/printer/RNNetPrinterModule.java [hunk 8]
```diff
@@ -58,8 +58,13 @@ public void connectPrinter(String host, Integer port, Callback successCallback,
     }
 
     @ReactMethod
-    public void connectAndSend(String host, Integer port, String data, Callback successCallback, Callback errorCallback) {
+    public void connectAndSend(String host, Integer port, String data, String brand, Callback successCallback, Callback errorCallback) {
+        // this.init(successCallback, errorCallback);
```
Comment (snehithv) [inline android/src/main/java/com/pinmi/react/printer/RNNetPrinterModule.java:62]: why is this commented?

Comment (mdsiddiq-till) [reply android/src/main/java/com/pinmi/react/printer/RNNetPrinterModule.java:62]: This should be removed. You can find similar line below it. which does initialization

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java [hunk 9]
```diff
@@ -188,9 +176,53 @@ public void run() {
...
+        if(this.mBluetoothDevice.getAddress().equals(blePrinterDeviceId.getInnerMacAddress()) && this.mBluetoothSocket != null){
+          Log.v(LOG_TAG, "do not need to reconnect");
+          printRawData(data, errorCallback);
+          successCallback.invoke(new BLEPrinterDevice(this.mBluetoothDevice).toRNWritableMap());
+          return;
+        }else{
+          closeConnectionIfExists();
+        }
+      }
+      Set<BluetoothDevice> pairedDevices = getBTAdapter().getBondedDevices();
```
Comment (snehithv) [inline android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:206]: what are all these lines of code doing?

I see we are already printing in Line  199

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java [hunk 10]
```diff
@@ -188,9 +176,53 @@ public void run() {
...
+            return;
+          }catch (IOException e){
+            e.printStackTrace();
+            errorCallback.invoke(e.getMessage());
+            return;
+          }
+        }
+      }
+      String errorText = "Can not find the specified printing device, please perform Bluetooth pairing in the system settings first.";
+      Toast.makeText(this.mContext, errorText, Toast.LENGTH_LONG).show();
```
Comment (snehithv) [inline android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:224]: why are we sending a toast message

File: android/src/main/java/com/pinmi/react/printer/adapter/PrinterAdapter.java [hunk 11]
```diff
@@ -30,4 +30,5 @@
 
     public void connectAndSend(String host, Integer port, String rawBase64Data, Callback successCallback, Callback errorCallback);
 
+    public void connectAndSend(PrinterDeviceId printerDeviceId, String data, Callback successCallback, Callback errorCallback);
```
Comment (snehithv) [inline android/src/main/java/com/pinmi/react/printer/adapter/PrinterAdapter.java:33]: what is PrinterDeviceId? Its a type ?

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java (hunk 9 above)
Comment (mdsiddiq-till) [reply android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:206]: These list all the available bluetooth devices and loop over them and find whichever matches

File: ios/adapters/epson/BLEPrinterEpsonAdapter.h (hunk 5 above)
Comment (snehithv) [reply ios/adapters/epson/BLEPrinterEpsonAdapter.h:11]: Then this logic should be replicated in NetPrinter also right?

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java [hunk 12]
```diff
@@ -188,9 +176,53 @@ public void run() {
...
+            successCallback.invoke(new BLEPrinterDevice(this.mBluetoothDevice).toRNWritableMap());
+            return;
+          }catch (IOException e){
+            e.printStackTrace();
+            errorCallback.invoke(e.getMessage());
+            return;
+          }
+        }
+      }
+      String errorText = "Can not find the specified printing device, please perform Bluetooth pairing in the system settings first.";
```
Comment (NalinSajwan) [inline android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:222]: Not planning on error enums (as we discussed during Dev collab) ? Or something to add later ?

File: ios/utils/NSDataAdditions.m [hunk 13]
```diff
@@ -83,4 +83,20 @@ - (id) initWithBase64EncodedString:(NSString *) string {
...
 
+- (NSString *)hexadecimalString {
+    /* Returns hexadecimal string of NSData. Empty string if data is empty.   */
+
+    const unsigned char *dataBuffer = (const unsigned char *)[self bytes];
+
+    if (!dataBuffer)
+        return [NSString string];
+
+    NSUInteger          dataLength  = [self length];
```
Comment (snehithv) [inline ios/utils/NSDataAdditions.m:94]: why so much spacing?

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java (hunk 9 above)
Comment (mdsiddiq-till) [reply android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:206]: The line 199 is inside if block. If already matching device is present

File: ios/utils/NSDataAdditions.m (hunk 13 above)
Comment (mdsiddiq-till) [reply ios/utils/NSDataAdditions.m:94]: It was xcode formatted code.

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java [hunk 14]
```diff
@@ -188,9 +176,53 @@ public void run() {
...
     }
-
+    
+    @Override
+    public void connectAndSend(String host, Integer port, String rawBase64Data, Callback successCallback, Callback errorCallback) {
+    }
 
     @Override
-    public void connectAndSend(String host, Integer port, String data, Callback successCallback, Callback errorCallback) {
+    public void connectAndSend(PrinterDeviceId printerDeviceId, String data, Callback successCallback, Callback errorCallback) {
```
Comment (NalinSajwan) [inline android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:184]: Same function names being declared ?  Above one is for network and this one for any other use case ?

Or by brand ?

File: android/src/main/java/com/pinmi/react/printer/adapter/PrinterAdapter.java (hunk 11 above)
Comment (mdsiddiq-till) [reply android/src/main/java/com/pinmi/react/printer/adapter/PrinterAdapter.java:33]: It is printer device object which holds printer info

File: example/ios/ThermalReceiptPrinterExample/Images.xcassets/AppIcon.appiconset/Contents.json [hunk 15]
```diff
@@ -2,37 +2,52 @@
...
+    },
+    {
+      "idiom" : "iphone",
+      "scale" : "2x",
+      "size" : "60x60"
+    },
+    {
+      "idiom" : "iphone",
+      "scale" : "3x",
+      "size" : "60x60"
```
Comment (NalinSajwan) [inline example/ios/ThermalReceiptPrinterExample/Images.xcassets/AppIcon.appiconset/Contents.json:41]: What does this helps in ? Asset icon sizes ?

We are repeating sizes for same scales `2x` and `3x`.

**Review by NalinSajwan — COMMENTED** [review]
Added few questions (might be silly) 🙂.

A lot of new things to read (code wise).

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java (hunk 14 above)
Comment (mdsiddiq-till) [reply android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:184]: Yes. function overriding. for net and bluetooth printing.

File: android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java (hunk 12 above)
Comment (mdsiddiq-till) [reply android/src/main/java/com/pinmi/react/printer/adapter/BLEPrinterAdapter.java:222]: Not doing now. There are so many things to touch and doesn't add value at this point.

File: example/ios/ThermalReceiptPrinterExample/Images.xcassets/AppIcon.appiconset/Contents.json (hunk 15 above)
Comment (mdsiddiq-till) [reply example/ios/ThermalReceiptPrinterExample/Images.xcassets/AppIcon.appiconset/Contents.json:41]: No idea about the details. This is for example app and autogenerated after some configs

File: ios/adapters/epson/BLEPrinterEpsonAdapter.h (hunk 5 above)
Comment (mdsiddiq-till) [reply ios/adapters/epson/BLEPrinterEpsonAdapter.h:11]: These files are same as before.
Might be some conditional imports are not needed.
But, I put the files same as they were before.

File: ios/adapters/epson/USBPrinterEpsonAdapter.h (hunk 2 above)
Comment (mdsiddiq-till) [reply ios/adapters/epson/USBPrinterEpsonAdapter.h:8]: It is the condition

**Pull request merged**